crates/turborepo-scm/src/hash_object.rs

```diff
@@ -0,0 +1,188 @@
+use std::{
+    io::{BufWriter, Read, Write},
+    panic,
+    process::{Command, Stdio},
+    thread,
+};
+
+use nom::{Finish, IResult};
+use turbopath::{AbsoluteSystemPathBuf, RelativeUnixPathBuf};
+
+use crate::{package_deps::GitHashes, read_git_error, wait_for_success, Error};
+
+pub(crate) fn hash_objects(
+    pkg_path: &AbsoluteSystemPathBuf,
+    to_hash: Vec<RelativeUnixPathBuf>,
+    pkg_prefix: &RelativeUnixPathBuf,
+    hashes: &mut GitHashes,
+) -> Result<(), Error> {
+    if to_hash.is_empty() {
+        return Ok(());
+    }
+    let mut git = Command::new("git")
+        .args(["hash-object", "--stdin-paths"])
+        .current_dir(pkg_path)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .stdin(Stdio::piped())
+        .spawn()?;
+
+    let stdout = git
+        .stdout
+        .as_mut()
+        .ok_or_else(|| Error::git_error("failed to get stdout for git hash-object"))?;
+    // We take, rather than borrow, stdin so that we can drop it and force the
+    // underlying file descriptor to close, signalling the end of input.
+    let stdin: std::process::ChildStdin = git
+        .stdin
+        .take()
+        .ok_or_else(|| Error::git_error("failed to get stdin for git hash-object"))?;
+    let mut stderr = git
+        .stderr
+        .take()
+        .ok_or_else(|| Error::git_error("failed to get stderr for git hash-object"))?;
+    let result = read_object_hashes(stdout, stdin, &to_hash, pkg_prefix, hashes);
+    if let Err(err) = result {
+        return Err(read_git_error(&mut stderr).unwrap_or(err));
+    }
```

Comment:
nit: `map_err` might work better here

```suggestion
    let result = read_object_hashes(stdout, stdin, &to_hash, pkg_prefix, hashes).map_err(|err| {
      read_git_error(&mut stderr).unwrap_or(err)
    })?;
```